src/dict.c

```diff
@@ -984,32 +986,33 @@ static int _dictExpandIfNeeded(dict *d)
     if (dictIsRehashing(d)) return DICT_OK;
 
     /* If the hash table is empty expand it to the initial size. */
-    if (d->ht[0].size == 0) return dictExpand(d, DICT_HT_INITIAL_SIZE);
+    if (DICTHT_SIZE(d->ht_size_exp[0]) == 0) return dictExpand(d, DICT_HT_INITIAL_SIZE);
 
     /* If we reached the 1:1 ratio, and we are allowed to resize the hash
      * table (global setting) or we should avoid it but the ratio between
      * elements/buckets is over the "safe" threshold, we resize doubling
      * the number of buckets. */
-    if (d->ht[0].used >= d->ht[0].size &&
+    if (d->ht_used[0] >= DICTHT_SIZE(d->ht_size_exp[0]) &&
         (dict_can_resize ||
-         d->ht[0].used/d->ht[0].size > dict_force_resize_ratio) &&
+         d->ht_used[0]/ DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio) &&
         dictTypeExpandAllowed(d))
     {
-        return dictExpand(d, d->ht[0].used + 1);
+        return dictExpand(d, d->ht_used[0] + 1);
     }
     return DICT_OK;
 }
 
+/* TODO: clz optimization */
```

Comment:
i suggested the CLZ opt twice:
https://github.com/redis/redis/pull/2382
https://github.com/redis/redis/pull/3833
and got rejected.

Salvatore believe that it makes the code less readable, and we did agree that due to the way it is used, there's no real impact on performance.

personally, i'm think it makes the code less readable

Reply:
Note that now we can use the instrinsic `__builtin_clzl` supported by both clang and gcc. It's definitely more efficient and arguably more readable than both antirez's and your code. In any case I used it here: https://github.com/yoav-steinberg/redis/blob/52596812e7e9b74f1ca63ad197ab30665198437b/src/server.c#L1748

Reply:
i think we may need a wrapper with the C fallback. not sure if redis isn't also used by some compiler that doesn't have that builtin.